# Segmenting and Clustering Neighborhoods in Toronto City

## Introduction

In this capstone project, we will use the Foursquare API to explore neighborhoods in Toronto City. We will use the **explore** function to get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters. We will use the *k*-means clustering algorithm to complete this task. Finally, we will use the Folium library to visualize the neighborhoods in Toronto City and their emerging clusters.

## Table of Contents<Br>

***1. Download and Explore Dataset***<Br>
***2. Explore Neighborhoods in Toronto City***<Br>
***3. Analyze Each Neighborhood***<Br>
***4. Cluster Neighborhoods***<Br>
***5. Examine Clusters***<Br>
***6. Inference***<Br>

***Download all the dependencies that are required.***

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
import matplotlib.pyplot as plt

# import k-means from clustering stage
from sklearn.cluster import KMeans

from sklearn.metrics import silhouette_score

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Fetching package metadata .............
Solving package specifications: .

# All requested packages already installed.
# packages in environment at /opt/conda/envs/DSX-Python35:
#
geopy                     1.17.0                     py_0    conda-forge
Fetching package metadata .............
Solving package specifications: .

# All requested packages already installed.
# packages in environment at /opt/conda/envs/DSX-Python35:
#
folium                    0.5.0                      py_0    conda-forge
Libraries imported.


## 1. Download and Explore Dataset

Neighborhood has a total of 11 boroughs and 212 neighborhoods. In order to segement the neighborhoods and explore them, we will essentially need a dataset that contains the 11 boroughs and the neighborhoods that exist in each borough as well as the latitude and logitude coordinates of each neighborhood.

***Load and explore the data***

In [2]:
Geo_data=pd.read_csv('Toronto_Geo.csv')
print('Data downloaded!')

Data downloaded!


In [3]:
Geo_data.head()

Unnamed: 0 PostalCode      Borough                            Neighborhood  \
0           0        M1B  Scarborough                          Rouge, Malvern   
1           1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union   
2           2        M1E  Scarborough       Guildwood, Morningside, West Hill   
3           3        M1G  Scarborough                                  Woburn   
4           4        M1H  Scarborough                               Cedarbrae   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

In [4]:
Geo_data = Geo_data.drop('Unnamed: 0', axis=1)
Geo_data.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [5]:
Geo_data = Geo_data[['PostalCode','Borough','Latitude','Longitude','Neighborhood']] #reorder columns
Geo_data.head()

PostalCode      Borough   Latitude  Longitude  \
0        M1B  Scarborough  43.806686 -79.194353   
1        M1C  Scarborough  43.784535 -79.160497   
2        M1E  Scarborough  43.763573 -79.188711   
3        M1G  Scarborough  43.770992 -79.216917   
4        M1H  Scarborough  43.773136 -79.239476   

                             Neighborhood  
0                          Rouge, Malvern  
1  Highland Creek, Rouge Hill, Port Union  
2       Guildwood, Morningside, West Hill  
3                                  Woburn  
4                               Cedarbrae

***Split Neighborhood column***

In [6]:
def explode(Geo_data, columns):
    idx = np.repeat(Geo_data.index, Geo_data[columns[0]].str.len())
    a = Geo_data.T.reindex(columns).values
    concat = np.concatenate([np.concatenate(a[i]) for i in range(a.shape[0])])
    p = pd.DataFrame(concat.reshape(a.shape[0], -1).T, idx, columns)
    return pd.concat([Geo_data.drop(columns, axis=1), p], axis=1).reset_index(drop=True)

In [7]:
Geo_data.Neighborhood = Geo_data.Neighborhood.str.split(',')

In [8]:
Geo_data=explode(Geo_data, ['Neighborhood'])

In [9]:
Geo_data.shape

(212, 5)

In [10]:
Geo_data

PostalCode           Borough   Latitude  Longitude  \
0          M1B       Scarborough  43.806686 -79.194353   
1          M1B       Scarborough  43.806686 -79.194353   
2          M1C       Scarborough  43.784535 -79.160497   
3          M1C       Scarborough  43.784535 -79.160497   
4          M1C       Scarborough  43.784535 -79.160497   
5          M1E       Scarborough  43.763573 -79.188711   
6          M1E       Scarborough  43.763573 -79.188711   
7          M1E       Scarborough  43.763573 -79.188711   
8          M1G       Scarborough  43.770992 -79.216917   
9          M1H       Scarborough  43.773136 -79.239476   
10         M1J       Scarborough  43.744734 -79.239476   
11         M1K       Scarborough  43.727929 -79.262029   
12         M1K       Scarborough  43.727929 -79.262029   
13         M1K       Scarborough  43.727929 -79.262029   
14         M1L       Scarborough  43.711112 -79.284577   
15         M1L       Scarborough  43.711112 -79.284577   
16         M1L       Scarborough  43.711112 -79.284577   
17         M1M       Scarborough  43.716316 -79.239476   
18         M1M       Scarborough  43.716316 -79.239476   
19         M1M       Scarborough  43.716316 -79.239476   
20         M1N       Scarborough  43.692657 -79.264848   
21         M1N       Scarborough  43.692657 -79.264848   
22         M1P       Scarborough  43.757410 -79.273304   
23         M1P       Scarborough  43.757410 -79.273304   
24         M1P       Scarborough  43.757410 -79.273304   
25         M1R       Scarborough  43.750071 -79.295849   
26         M1R       Scarborough  43.750071 -79.295849   
27         M1S       Scarborough  43.794200 -79.262029   
28         M1T       Scarborough  43.781638 -79.304302   
29         M1T       Scarborough  43.781638 -79.304302   
30         M1T       Scarborough  43.781638 -79.304302   
31         M1V       Scarborough  43.815252 -79.284577   
32         M1V       Scarborough  43.815252 -79.284577   
33         M1V       Scarborough  43.815252 -79.284577   
34         M1V       Scarborough  43.815252 -79.284577   
35         M1W       Scarborough  43.799525 -79.318389   
36         M1W       Scarborough  43.799525 -79.318389   
37         M1X       Scarborough  43.836125 -79.205636   
38         M2H        North York  43.803762 -79.363452   
39         M2J        North York  43.778517 -79.346556   
40         M2J        North York  43.778517 -79.346556   
41         M2J        North York  43.778517 -79.346556   
42         M2K        North York  43.786947 -79.385975   
43         M2L        North York  43.757490 -79.374714   
44         M2L        North York  43.757490 -79.374714   
45         M2M        North York  43.789053 -79.408493   
46         M2M        North York  43.789053 -79.408493   
47         M2N        North York  43.770120 -79.408493   
48         M2P        North York  43.752758 -79.400049   
49         M2R        North York  43.782736 -79.442259   
50         M3A        North York  43.753259 -79.329656   
51         M3B        North York  43.745906 -79.352188   
52         M3C        North York  43.725900 -79.340923   
53         M3C        North York  43.725900 -79.340923   
54         M3H        North York  43.754328 -79.442259   
55         M3H        North York  43.754328 -79.442259   
56         M3H        North York  43.754328 -79.442259   
57         M3J        North York  43.767980 -79.487262   
58         M3J        North York  43.767980 -79.487262   
59         M3K        North York  43.737473 -79.464763   
60         M3K        North York  43.737473 -79.464763   
61         M3L        North York  43.739015 -79.506944   
62         M3M        North York  43.728496 -79.495697   
63         M3N        North York  43.761631 -79.520999   
64         M4A        North York  43.725882 -79.315572   
65         M4B         East York  43.706397 -79.309937   
66         M4B         East York  43.706397 -79.309937   
67         M4C         East York  43.695344 -79.318389   
68

In [11]:
Geo_data.isnull().sum()

PostalCode      0
Borough         0
Latitude        0
Longitude       0
Neighborhood    0
dtype: int64

***Use geopy library to get the latitude and longitude values of Toronto City. The geograpical coordinate of Toronto are 43.653963, -79.387207***

In [12]:
import time
start_time = time.time()

address = 'Toronto, ON'

geolocator = Nominatim(user_agent="CanadaApp")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

print("--- %s seconds ---" % round((time.time() - start_time), 2))

The geograpical coordinate of Toronto City are 43.653963, -79.387207.
--- 0.29 seconds ---


***Create a map of Toronto with neighborhoods superimposed on top.***

In [13]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(Geo_data['Latitude'], Geo_data['Longitude'], Geo_data['Borough'], Geo_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

***The dataset has all 11 boroughs and 212 neighborhoods.***

In [14]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(Geo_data['Borough'].unique()),
        len(Geo_data['Neighborhood'].unique())))
print(Geo_data.shape)

The dataframe has 11 boroughs and 212 neighborhoods.
(212, 5)


***Define Foursquare Credentials and Version***

In [15]:
CLIENT_ID = 'OLN1BAQQBHO234LKFIU1ZNGV4Z3O3P1GS5KIMTNPJHLX1MKL' # your Foursquare ID
CLIENT_SECRET = 'VDM5CGGVSUOGKMY21ETO4J1UAJH5QJEALQCJAIWUF2DJXR2T' # your Foursquare Secret
VERSION = '20181120' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: OLN1BAQQBHO234LKFIU1ZNGV4Z3O3P1GS5KIMTNPJHLX1MKL
CLIENT_SECRET:VDM5CGGVSUOGKMY21ETO4J1UAJH5QJEALQCJAIWUF2DJXR2T


***Get the first neighborhood's name.***

In [16]:
Geo_data.loc[0, 'Neighborhood']

'Rouge'

***Get the neighborhood's latitude and longitude values.***

In [17]:
neighborhood_latitude = Geo_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = Geo_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = Geo_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Rouge are 43.8066863, -79.19435340000003.


***Get the top 100 venues that are in Rouge within a radius of 500 meters.***

In [18]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=OLN1BAQQBHO234LKFIU1ZNGV4Z3O3P1GS5KIMTNPJHLX1MKL&client_secret=VDM5CGGVSUOGKMY21ETO4J1UAJH5QJEALQCJAIWUF2DJXR2T&v=20181120&ll=43.8066863,-79.19435340000003&radius=500&limit=100'

***Send the GET request and examine the resutls***

In [19]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5c00dc054434b95331af6d5d'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-4bb6b9446edc76b0d771311c-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/fastfood_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d16e941735',
         'name': 'Fast Food Restaurant',
         'pluralName': 'Fast Food Restaurants',
         'primary': True,
         'shortName': 'Fast Food'}],
       'id': '4bb6b9446edc76b0d771311c',
       'location': {'cc': 'CA',
        'city': 'Toronto',
        'country': 'Canada',
        'crossStreet': 'Morningside & Sheppard',
        'distance': 387,
        'formattedAddress': ['Toronto ON', 'Canada'],
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.80744841934756,
          'ln

In [20]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

***Clean the json and structure it into a pandas dataframe.***

In [21]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat        lng
0  Wendy's  Fast Food Restaurant  43.807448 -79.199056

In [22]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

1 venues were returned by Foursquare.


## 2. Explore Neighborhoods in Toronto City

#### Let's create a function to repeat the same process to all the neighborhoods in Toronto

In [23]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

***Run the above function on each neighborhood and create a new dataframe called Toronto_venues.***

In [24]:
Toronto_venues = getNearbyVenues(names=Geo_data['Neighborhood'],
                                  latitudes=Geo_data['Latitude'],
                                  longitudes=Geo_data['Longitude']
                                  )

Rouge
 Malvern
Highland Creek
 Rouge Hill
 Port Union
Guildwood
 Morningside
 West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park
 Ionview
 Kennedy Park
Clairlea
 Golden Mile
 Oakridge
Cliffcrest
 Cliffside
 Scarborough Village West
Birch Cliff
 Cliffside West
Dorset Park
 Scarborough Town Centre
 Wexford Heights
Maryvale
 Wexford
Agincourt
Clarks Corners
 Sullivan
 Tam O'Shanter
Agincourt North
 L'Amoreaux East
 Milliken
 Steeles East
L'Amoreaux West
 Steeles West
Upper Rouge
Hillcrest Village
Fairview
 Henry Farm
 Oriole
Bayview Village
Silver Hills
 York Mills
Newtonbrook
 Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park
 Don Mills South
Bathurst Manor
 Downsview North
 Wilson Heights
Northwood Park
 York University
CFB Toronto
 Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens
 Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The D

#### Let's check the size of the resulting dataframe

In [25]:
print(Toronto_venues.shape)
Toronto_venues.head()

(4456, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0           Rouge              43.806686              -79.194353   
1         Malvern              43.806686              -79.194353   
2  Highland Creek              43.784535              -79.160497   
3  Highland Creek              43.784535              -79.160497   
4      Rouge Hill              43.784535              -79.160497   

                    Venue  Venue Latitude  Venue Longitude  \
0                 Wendy's       43.807448       -79.199056   
1                 Wendy's       43.807448       -79.199056   
2  Chris Effects Painting       43.784343       -79.163742   
3   Royal Canadian Legion       43.782533       -79.163085   
4  Chris Effects Painting       43.784343       -79.163742   

               Venue Category  
0        Fast Food Restaurant  
1        Fast Food Restaurant  
2  Construction & Landscaping  
3                         Bar  
4  Construction & Landscaping

***Let's check how many venues were returned for each neighborhood***

In [26]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                              
 Bathurst Quay                                                       13   
 Beaumond Heights                                                    13   
 Cliffside                                                            3   
 Cliffside West                                                       4   
 Don Mills South                                                     20   
 Downsview East                                                       4   
 Downsview North                                                     18   
 Dufferin                                                            18   
 Eringate                                                             7   
 Exhibition Place                                                    22   
 Forest Hill SE                                                      14   
 Forest Hill West                                                     4   
 Garden District                                                    100   
 Golden Mile                                                          9   
 Grange Park                                                        100   
 Harbourfront West                                                   13   
 Henry Farm                                                          64   
 Humbergate                                                          13   
 Humberlea                                                            1   
 India Bazaar                                                        19   
 Ionview                                                              7   
 Island airport                                                      13   
 Islington                                                            1   
 Jamestown                                                           13   
 Keelesdale                                                           4   
 Kennedy Park                                                         7   
 Kensington Market                                                  100   
 King                                                               100   
 King and Spadina                                                    13   
 King's Mill Park                                                     1   
 Kingsway Park South East                                             1   
 L'Amoreaux East                                                      3   
 Lawrence Manor                                                      12   
 Lawrence Manor East                                                 25   
 Long Branch                                                         11   
 Malvern                                                              1   
 Markland Wood                                                        7   
 Martin Grove                                                         1   
 Martin Grove Gardens                                                 2   
 Milliken                                                             3   
 Mimico NE                                                            1   
 Mimico NW                                                           14   
 Mimico South                                                        16   
 Montgomery Road                                                      2   
 Morningside                                                          6   
 Mount Dennis                                                         4   
 Mount Olive                                                         13   
 New Toronto                                                         16   
 North Midtown                                                       24   
 North Park                                                           4   
 Oakridge                                                             9   
 Old Burnhamthorpe                                                    7   


***Let's find out how many unique categories can be curated from all the returned venues***

In [27]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 273 uniques categories.


## 3. Analyze Each Neighborhood

In [28]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0            0                  0               0                  0        0   
1            0                  0               0                  0        0   
2            0                  0               0                  0        0   
3            0                  0               0                  0        0   
4            0                  0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0                 0                    0             0         0   
1                 0                    0             0         0   
2                 0                    0             0         0   
3                 0                    0             0         0   
4                 0                    0             0         0   

   Arepa Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                 0            0           0                    0   
1                 0            0           0                    0   
2                 0            0           0                    0   
3                 0            0           0                    0   
4                 0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0                 0                   0            0              0   
1                 0                   0            0              0   
2                 0                   0            0              0   
3                 0                   0            0              0   
4                 0                   0            0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    1               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  Beer Bar  \
0                 0                 0                   0      0         0   
1                 0                 0                   0      0         0   
2                 0                 0                   0      0         0   
3                 0                 0                   0      0         0   
4                 0                 0                   0      0         0   

   Beer Store  Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  \
0           0                   0          0       0              0   
1           0                   0          0       0              0   
2           0                   0          0       0              0   
3           0                   0          0       0              0   
4           0                   0          0       0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bridal Shop  Bubble Tea Shop  Building  Bur

***And let's examine the new dataframe size.***

In [29]:
Toronto_onehot.shape

(4456, 273)

In [30]:
# Reorder Columns
cols = Toronto_onehot.columns.tolist()
n = int(cols.index('Neighborhood'))
cols = [cols[n]] + cols[:n] + cols[n+1:]
Toronto_onehot = Toronto_onehot[cols]

In [31]:
Toronto_onehot.head()

Neighborhood  Yoga Studio  Accessories Store  Adult Boutique  \
0           Rouge            0                  0               0   
1         Malvern            0                  0               0   
2  Highland Creek            0                  0               0   
3  Highland Creek            0                  0               0   
4      Rouge Hill            0                  0               0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Arepa Restaurant  Art Gallery  Art Museum  \
0             0         0                 0            0           0   
1             0         0                 0            0           0   
2             0         0                 0            0           0   
3             0         0                 0            0           0   
4             0         0                 0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    1   
4              0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0               0                 0                 0                   0   
1               0                 0                 0                   0   
2               0                 0                 0                   0   
3               0                 0                 0                   0   
4               0                 0                 0                   0   

   Beach  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  Bistro  \
0      0         0           0                   0          0       0   
1      0         0           0                   0          0       0   
2      0         0           0                   0          0       0   
3      0         0           0                   0          0       0   
4      0         0           0                   0          0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Bre

In [32]:
Toronto_onehot.shape

(4456, 273)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [33]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()
Toronto_grouped

Neighborhood  Yoga Studio  \
0                                       Bathurst Quay     0.000000   
1                                    Beaumond Heights     0.000000   
2                                           Cliffside     0.000000   
3                                      Cliffside West     0.000000   
4                                     Don Mills South     0.000000   
5                                      Downsview East     0.000000   
6                                     Downsview North     0.000000   
7                                            Dufferin     0.000000   
8                                            Eringate     0.000000   
9                                    Exhibition Place     0.000000   
10                                     Forest Hill SE     0.000000   
11                                   Forest Hill West     0.000000   
12                                    Garden District     0.000000   
13                                        Golden Mile     0.000000   
14                                        Grange Park     0.000000   
15                                  Harbourfront West     0.000000   
16                                         Henry Farm     0.000000   
17                                         Humbergate     0.000000   
18                                          Humberlea     0.000000   
19                                       India Bazaar     0.000000   
20                                            Ionview     0.000000   
21                                     Island airport     0.000000   
22                                          Islington     0.000000   
23                                          Jamestown     0.000000   
24                                         Keelesdale     0.000000   
25                                       Kennedy Park     0.000000   
26                                  Kensington Market     0.000000   
27                                               King     0.000000   
28                                   King and Spadina     0.000000   
29                                   King's Mill Park     0.000000   
30                           Kingsway Park South East     0.000000   
31                                    L'Amoreaux East     0.000000   
32                                     Lawrence Manor     0.000000   
33                                Lawrence Manor East     0.000000   
34                                        Long Branch     0.000000   
35                                            Malvern     0.000000   
36                                      Markland Wood     0.000000   
37                                       Martin Grove     0.000000   
38                               Martin Grove Gardens     0.000000   
39                                           Milliken     0.000000   
40                                          Mimico NE     0.000000   
41                                          Mimico NW     0.000000   
42                                       Mimico South     0.000000   
43                                    Montgomery Road     0.000000   
44                                        Morningside     0.000000   
45                                       Mount Dennis     0.000000   
46                                        Mount Olive     0.000000   
47                                        New Toronto     0.000000   
48                                      North Midtown     0.000000   
49                                         North Park     0.000000   
50                                           Oakridge     0.000000   
51                                  Old Burnhamthorpe     0.000000   
52                                     Old Mill North     0.000000   
53                                     Old Mill South     0.000000   
54                                             Oriole     0.000000   
55                                   Parkdale Village     0.000000   
56                                      Parkview Hi

#### Let's confirm the new size

In [34]:
Toronto_grouped.shape

(208, 273)

#### Let's print each neighborhood along with the top 5 most common venues

In [35]:
num_top_venues = 5

for hood in Toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

---- Bathurst Quay----
              venue  freq
0    Airport Lounge  0.15
1   Airport Service  0.15
2  Airport Terminal  0.15
3     Boat or Ferry  0.08
4          Boutique  0.08


---- Beaumond Heights----
                 venue  freq
0          Pizza Place  0.15
1        Grocery Store  0.15
2       Discount Store  0.08
3          Video Store  0.08
4  Fried Chicken Joint  0.08


---- Cliffside----
                 venue  freq
0         Intersection  0.33
1                Motel  0.33
2  American Restaurant  0.33
3        Movie Theater  0.00
4  Monument / Landmark  0.00


---- Cliffside West----
                   venue  freq
0        College Stadium  0.25
1           Skating Rink  0.25
2  General Entertainment  0.25
3                   Café  0.25
4     Miscellaneous Shop  0.00


---- Don Mills South----
              venue  freq
0       Coffee Shop  0.10
1               Gym  0.10
2        Beer Store  0.10
3  Asian Restaurant  0.10
4    Clothing Store  0.05


---- Downsview East----
   

#### Let's put that into a *pandas* dataframe

***First, let's write a function to sort the venues in descending order.***

In [36]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

***Now let's create the new dataframe and display the top 10 venues for each neighborhood.***

In [37]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood  \
0                                       Bathurst Quay   
1                                    Beaumond Heights   
2                                           Cliffside   
3                                      Cliffside West   
4                                     Don Mills South   
5                                      Downsview East   
6                                     Downsview North   
7                                            Dufferin   
8                                            Eringate   
9                                    Exhibition Place   
10                                     Forest Hill SE   
11                                   Forest Hill West   
12                                    Garden District   
13                                        Golden Mile   
14                                        Grange Park   
15                                  Harbourfront West   
16                                         Henry Farm   
17                                         Humbergate   
18                                          Humberlea   
19                                       India Bazaar   
20                                            Ionview   
21                                     Island airport   
22                                          Islington   
23                                          Jamestown   
24                                         Keelesdale   
25                                       Kennedy Park   
26                                  Kensington Market   
27                                               King   
28                                   King and Spadina   
29                                   King's Mill Park   
30                           Kingsway Park South East   
31                                    L'Amoreaux East   
32                                     Lawrence Manor   
33                                Lawrence Manor East   
34                                        Long Branch   
35                                            Malvern   
36                                      Markland Wood   
37                                       Martin Grove   
38                               Martin Grove Gardens   
39                                           Milliken   
40                                          Mimico NE   
41                                          Mimico NW   
42                                       Mimico South   
43                                    Montgomery Road   
44                                        Morningside   
45                                       Mount Dennis   
46                                        Mount Olive   
47                                        New Toronto   
48                                      North Midtown   
49                                         North Park   
50                                           Oakridge   
51                                  Old Burnhamthorpe   
52                                     Old Mill North   
53                                     Old Mill South   
54                                             Oriole   
55                                   Parkdale Village   
56                                      Parkview Hill   
57                                         Port Union   
58                                   Princess Gardens   
59                                      Railway Lands   
60                                          Rathnelly   
61                                        Regent Park   
62                                           Richmond   
63                                   Richview Gardens   
64                                          Riverdale   
65                                       Roncesvalles   
66                                         Rouge Hill   
67                              Royal York South East   
68                              Royal York South West   
69                                          Runnyme

## 4. Cluster Neighborhoods

***First, let's determine the optimal value of K for our dataset using the Silhouette Coefficient Method***

***From sklearn documentation - https://scikit-learn.org/stable/modules/clustering.html#silhouette-coefficient***

A higher Silhouette Coefficient score relates to a model with better-defined clusters. The Silhouette Coefficient is defined for each sample and is composed of two scores: `

    a: The mean distance between a sample and all other points in the same class.

    b: The mean distance between a sample and all other points in the next nearest cluster.

The Silhouette Coefficient is for a single sample is then given as:

s=b-a/max(a,b)

Now, to find the optimal value of k for KMeans, loop through 1..n for n_clusters in KMeans and calculate Silhouette Coefficient for each sample.

A higher Silhouette Coefficient indicates that the object is well matched to its own cluster and poorly matched to neighboring clusters.

In [38]:
Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

for n_cluster in range(2, 26):
    kmeans = KMeans(n_clusters=n_cluster).fit(Toronto_grouped_clustering)
    label = kmeans.labels_
    sil_coeff = silhouette_score(Toronto_grouped_clustering, label, metric='euclidean')
    print("For n_clusters={}, The Silhouette Coefficient is {}".format(n_cluster, sil_coeff))

For n_clusters=2, The Silhouette Coefficient is 0.508749867720038
For n_clusters=3, The Silhouette Coefficient is 0.5355438707701862
For n_clusters=4, The Silhouette Coefficient is 0.3994981831882502
For n_clusters=5, The Silhouette Coefficient is 0.4144931348929341
For n_clusters=6, The Silhouette Coefficient is 0.3864601857985886
For n_clusters=7, The Silhouette Coefficient is 0.3619720339509345
For n_clusters=8, The Silhouette Coefficient is 0.20854017049654328
For n_clusters=9, The Silhouette Coefficient is 0.22287745661442585
For n_clusters=10, The Silhouette Coefficient is 0.2439231914091651
For n_clusters=11, The Silhouette Coefficient is 0.262837284406974
For n_clusters=12, The Silhouette Coefficient is 0.25473691384639074
For n_clusters=13, The Silhouette Coefficient is 0.2845535102000928
For n_clusters=14, The Silhouette Coefficient is 0.28079392528453545
For n_clusters=15, The Silhouette Coefficient is 0.29173087998923
For n_clusters=16, The Silhouette Coefficient is 0.31531

***As we can see, n_clusters=3 has highest Silhouette Coefficient. This means that 3 should be the optimal number of clusters.<br>
For n_clusters=3, The Silhouette Coefficient is 0.5355438707701862***

***Run k-means to cluster the neighborhood into 3 clusters.***

In [39]:
# set number of clusters
kclusters = 3

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 1, 0, 0, 0, 0], dtype=int32)

***Now let's put our cluster labels in our neighborhoods_venues_sorted df***

***First, we need to remove the rows in our original df for which the Foursquare call didn't return any venues***

In [43]:
Geo_data['Neighborhood'] = neighborhoods_venues_sorted['Neighborhood']
print(Geo_data.shape)
Geo_data.isnull().sum()
Geo_data.dropna(how = 'any', axis = 0, inplace = True)
print(Geo_data.shape)
print(neighborhoods_venues_sorted.shape)

(212, 5)
(208, 5)
(208, 11)


***Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.***

In [55]:
Toronto_merged = Geo_data

# add clustering labels
Toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Toronto_merged.head() # check the last columns!

PostalCode      Borough   Latitude  Longitude       Neighborhood  \
0        M1B  Scarborough  43.806686 -79.194353      Bathurst Quay   
1        M1B  Scarborough  43.806686 -79.194353   Beaumond Heights   
2        M1C  Scarborough  43.784535 -79.160497          Cliffside   
3        M1C  Scarborough  43.784535 -79.160497     Cliffside West   
4        M1C  Scarborough  43.784535 -79.160497    Don Mills South   

   Cluster Labels  1st Most Common Venue 2nd Most Common Venue  \
0               0       Airport Terminal       Airport Service   
1               0          Grocery Store           Pizza Place   
2               0                  Motel          Intersection   
3               0  General Entertainment       College Stadium   
4               0       Asian Restaurant                   Gym   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0        Airport Lounge       Harbor / Marina          Airport Gate   
1            Beer Store   Fried Chicken Joint   Japanese Restaurant   
2   American Restaurant          Dessert Shop                 Diner   
3          Skating Rink                  Café         Women's Store   
4           Coffee Shop            Beer Store         Grocery Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0      Sculpture Garden              Boutique         Boat or Ferry   
1  Fast Food Restaurant        Discount Store              Pharmacy   
2        Discount Store               Dog Run      Doner Restaurant   
3                 Diner        Discount Store               Dog Run   
4        Clothing Store    Chinese Restaurant    Dim Sum Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0    Airport Food Court                Airport  
1        Sandwich Place            Coffee Shop  
2            Donut Shop              Drugstore  
3      Doner Restaurant             Donut Shop  
4            Restaurant         Sandwich Place

***Finally, let's visualize the resulting clusters***

In [56]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighborhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters

***Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster.***

#### Cluster 1

In [57]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
0         Scarborough               0           Airport Terminal   
1         Scarborough               0              Grocery Store   
2         Scarborough               0                      Motel   
3         Scarborough               0      General Entertainment   
4         Scarborough               0           Asian Restaurant   
6         Scarborough               0                Coffee Shop   
7         Scarborough               0                     Bakery   
8         Scarborough               0               Liquor Store   
9         Scarborough               0                Coffee Shop   
10        Scarborough               0                Coffee Shop   
12        Scarborough               0                Coffee Shop   
13        Scarborough               0                   Bus Line   
14        Scarborough               0                       Café   
15        Scarborough               0           Airport Terminal   
16        Scarborough               0             Clothing Store   
17        Scarborough               0              Grocery Store   
19        Scarborough               0                       Park   
20        Scarborough               0              Train Station   
21        Scarborough               0           Airport Terminal   
22        Scarborough               0                       Bank   
23        Scarborough               0              Grocery Store   
24        Scarborough               0                 Restaurant   
25        Scarborough               0              Train Station   
26        Scarborough               0                       Café   
27        Scarborough               0                Coffee Shop   
28        Scarborough               0           Airport Terminal   
32        Scarborough               0     Furniture / Home Store   
33        Scarborough               0         Italian Restaurant   
34        Scarborough               0                Pizza Place   
35        Scarborough               0       Fast Food Restaurant   
36        Scarborough               0               Liquor Store   
37        Scarborough               0                       Bank   
41         North York               0              Grocery Store   
42         North York               0                Coffee Shop   
44         North York               0        Rental Car Location   
45         North York               0                 Restaurant   
46         North York               0              Grocery Store   
47         North York               0                Coffee Shop   
48         North York               0                Coffee Shop   
50         North York               0                   Bus Line   
51         North York               0               Liquor Store   
54         North York               0             Clothing Store   
55         North York               0                Coffee Shop   
56         North York               0       Fast Food Restaurant   
57         North York               0                        Bar   
58         North York               0                       Bank   
59         North York               0           Airport Terminal   
60         North York               0                Coffee Shop   
61         North York               0                Coffee Shop   
62         North York               0                Coffee Shop   
64         North York               0           Greek Restaurant   
65          East York               0             Breakfast Spot   
66          East York               0                        Bar   
68       East Toronto               0              Grocery Store   
69          East York               0                   Bus Line   
70          East York               0          Indian Restaurant   
71          East York               0                      Motel   
72       East Toronto               0              Grocery Store   
73

#### Cluster 2

In [58]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
5         Scarborough               1                  Park   
11        Scarborough               1                  Park   
31        Scarborough               1                  Park   
38         North York               1                  Park   
39         North York               1                  Park   
43         North York               1                 River   
49         North York               1                  Park   
52         North York               1                 River   
63         North York               1                  Park   
79    Central Toronto               1                  Park   
80    Central Toronto               1                  Park   
83    Central Toronto               1                   Gym   
98   Downtown Toronto               1                  Park   
110        North York               1                  Park   
121  Downtown Toronto               1                  Park   
124  Downtown Toronto               1                  Park   
148              York               1                  Park   
153              York               1                  Park   
166         Etobicoke               1                  Park   
170         Etobicoke               1                  Park   
173         Etobicoke               1                  Park   
175         Etobicoke               1                   Gym   
180         Etobicoke               1  Fast Food Restaurant   
182         Etobicoke               1                  Park   
197              York               1                 River   
201         Etobicoke               1                  Park   
207         Etobicoke               1                  Park   

          2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
5                       Airport     Electronics Store              Bus Stop   
11             Sushi Restaurant         Jewelry Store                 Trail   
31                   Playground      Sculpture Garden            Donut Shop   
38                  Pizza Place         Women's Store             Drugstore   
39                   Playground      Sculpture Garden            Donut Shop   
43                         Park             Drugstore    Dim Sum Restaurant   
49   Construction & Landscaping      Basketball Court                Bakery   
52                         Park             Drugstore    Dim Sum Restaurant   
63                  Pizza Place         Women's Store             Drugstore   
79                  Pizza Place         Women's Store             Drugstore   
80                   Playground      Sculpture Garden            Donut Shop   
83                        Trail            Playground          Intersection   
98   Construction & Landscaping      Basketball Court                Bakery   
110                  Playground      Sculpture Garden            Donut Shop   
121                     Airport     Electronics Store              Bus Stop   
124               Women's Store              Pharmacy  Fast Food Restaurant   
148                 Coffee Shop     Convenience Store         Women's Store   
153            Sushi Restaurant         Jewelry Store                 Trail   
166                 Pizza Place         Women's Store             Drugstore   
170          Dim Sum Restaurant           Swim School              Bus Line   
173  Construction & Landscaping      Basketball Court                Bakery   
175                       Trail            Playground          Intersection   
180           Food & Drink Shop              Bus Stop                  Park   
182                  Playground                 Trail            Donut Shop   
197                        Park             Drugstore    Dim Sum Restaurant   
201               Women's Store   Dumpling Restaurant                 Diner   
207           Electronics Store                  Bank         Women's Store   

       5th Most Common Venue 6th Mos

#### Cluster 3

In [60]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18        Scarborough               2        Baseball Field   
29        Scarborough               2        Baseball Field   
30        Scarborough               2        Baseball Field   
40         North York               2        Baseball Field   
53         North York               2        Baseball Field   
67          East York               2        Baseball Field   
85    Central Toronto               2        Baseball Field   
89   Downtown Toronto               2        Baseball Field   
144      West Toronto               2        Baseball Field   
149              York               2        Baseball Field   
162      East Toronto               2        Baseball Field   

    2nd Most Common Venue        3rd Most Common Venue  \
18          Women's Store  Eastern European Restaurant   
29          Women's Store  Eastern European Restaurant   
30          Women's Store  Eastern European Restaurant   
40          Women's Store  Eastern European Restaurant   
53          Women's Store  Eastern European Restaurant   
67          Women's Store  Eastern European Restaurant   
85          Women's Store  Eastern European Restaurant   
89          Women's Store  Eastern European Restaurant   
144            Food Truck                Women's Store   
149         Women's Store  Eastern European Restaurant   
162         Women's Store  Eastern European Restaurant   

           4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
18                Discount Store               Dog Run      Doner Restaurant   
29                Discount Store               Dog Run      Doner Restaurant   
30                Discount Store               Dog Run      Doner Restaurant   
40                Discount Store               Dog Run      Doner Restaurant   
53                Discount Store               Dog Run      Doner Restaurant   
67                Discount Store               Dog Run      Doner Restaurant   
85                Discount Store               Dog Run      Doner Restaurant   
89                Discount Store               Dog Run      Doner Restaurant   
144  Eastern European Restaurant               Dog Run      Doner Restaurant   
149               Discount Store               Dog Run      Doner Restaurant   
162               Discount Store               Dog Run      Doner Restaurant   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
18             Donut Shop             Drugstore   Dumpling Restaurant   
29             Donut Shop             Drugstore   Dumpling Restaurant   
30             Donut Shop             Drugstore   Dumpling Restaurant   
40             Donut Shop             Drugstore   Dumpling Restaurant   
53             Donut Shop             Drugstore   Dumpling Restaurant   
67             Donut Shop             Drugstore   Dumpling Restaurant   
85             Donut Shop             Drugstore   Dumpling Restaurant   
89             Donut Shop             Drugstore   Dumpling Restaurant   
144            Donut Shop             Drugstore   Dumpling Restaurant   
149            Donut Shop             Drugstore   Dumpling Restaurant   
162            Donut Shop             Drugstore   Dumpling Restaurant   

    10th Most Common Venue  
18       Electronics Store  
29       Electronics Store  
30       Electronics Store  
40       Electronics Store  
53       Electronics Store  
67       Electronics Store  
85       Electronics Store  
89       Electronics Store  
144      Electronics Store  
149      Electronics Store  
162      Electronics Store

## 6. Inference

### Cluster 1 : Commercial Cluster<Br>
***Travel Modes -*** Airport Terminal, Rental Car Location, Train Station, Bus Line.<Br>
***Importantnt Travel Locations -*** Bank,Grocery Store,Shopping Mall<Br>
***Entertainment -*** General Entertainment, Performing Arts Venue<Br>
***Food Joints -*** Coffee Shops , Liquor Store<Br>
***Food Joints serving cuisine of various countries -*** Indian Restaurant, Latin American Restaurant,American Restaurant,Vietnamese Restaurant,<Br>
Chinese Restaurant,Middle Eastern Restaurant

### Cluster 2 : Zen Cluster <Br>
Park and Playground are the two most common venues.<Br> 
Fast Food Restaurants like Pizza Place, Dim Sum Restaurant, Food & Drink Shop, Dumpling Restaurant.

### Cluster 3 : Exclusive Cluster<Br>
Baseball Field, Women's Store  and Eastern European Restaurant are the three most common venues in this cluster.